osemgrep: enable tests where there are no rules

hannesm opened the pull request:
Two more tests are enabled, one is fixed (badlanguage.yaml file location); also a new one (does_not_exist.yaml) has been added.

osemgrep now exits with error code 7.

test plan:
 - make osempass

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

hannesm committed:
osemgrep: raise missing_configuration (exit code 7) when there are no rules

this enables two more tests, test_regex_with_bad_language and test_rule_parser__empty

For the empty.yaml, instead of asserting false when yaml_to_generic returned
something that is not a ExprStmt, return the empty list of rules.

hannesm committed:
fix bad_language test (file is present in syntax subdirectory); add a non-existing rule file test - and fix osemgrep

hannesm committed:
pre-commit hook: add two empty lines between tests

aryx commented:
need to resolve conflict @hannesm

hannesm commented:
resolved conflict

aryx commented:
one regression in python, weird:
FAILED tests/e2e/test_rule_parser.py::test_regex_with_bad_language - assert 8...

hannesm committed:
remove osempass from test_regex_with_bad_language for now, and adjust to exit code 8

at the moment, the exit code 8 isn't returned by osemgrep (since all errors are
passed through Core_runner)

hannesm committed:
comment about the test not passing

aryx commented on cli/tests/e2e/test_rule_parser.py:
```diff
@@ -31,11 +31,21 @@ def test_rule_parser__failure(run_semgrep_in_tmp: RunSemgrep, snapshot, filename
 
+@pytest.mark.osempass
+@pytest.mark.kinda_slow
+def test_nonexisting_faile(run_semgrep_in_tmp: RunSemgrep, snapshot):
```
typo, file, not faile

aryx commented on src/parsing/Parse_rule.ml:
```diff
@@ -1476,7 +1476,7 @@ let parse_generic_ast ?(error_recovery = false) (file : Fpath.t)
             yaml_error (Tok.tok_of_loc loc)
               "missing rules entry as top-level key")
-    | _ -> assert false
+    | _ -> (Tok.(tok_of_loc (first_loc_of_file !!file)), [])
```
I'm confused by this case. When does this happen? When the file does not contain anything?
Maybe we should pattern match more precisely on it? and assert false for the rest?
I'm scared we silently skip bad content or are tolerant to bad content (at least the assert false was returning
an error for bad stuff).
Or maybe it's ok because anyway the error will be detected later because we have empty rules? Then maybe
we need a comment there to reinsure the reader the issue will be detected at some point.

hannesm committed:
fix typo

hannesm committed:
only return the empty list for an empty file, as suggested by @aryx
